// users/routes/usersRoutes.js
const express = require('express');
const router = express.Router();
const User = require("../models/users");
const bcrypt = require('bcrypt');
const passport = require('passport');
const Helper = require("../helper/data");
const {ensureAuthenticated} = require("../../../config/auth.js")
const userController = require('../controllers/userController');


//login handle
router.get('/login',userController.userLoginPage);



router.get('/register',userController.userRegisterPage);




//Register handle
router.post('/login',(req,res,next)=>{
passport.authenticate('local',{
    successRedirect : '/users/dashboard',
    failureRedirect: '/users/login',
    failureFlash : true
})(req,res,next)
})






  //register post handle
  router.post('/register',(req,res)=>{
    const {name,email, password, password2} = req.body;
    let errors = [];
    console.log(' Name ' + name+ ' email :' + email+ ' pass:' + password);
    if(!name || !email || !password || !password2) {
        errors.push({msg : "Please fill in all fields"})
    }
    //check if match
    if(password !== password2) {
        errors.push({msg : "passwords dont match"});
    }
    
    //check if password is more than 6 characters
    if(password.length < 6 ) {
        errors.push({msg : 'password atleast 6 characters'})
    }
    if(errors.length > 0 ) {
    res.render(Helper.findView('register'), {
        errors : errors,
        name : name,
        email : email,
        password : password,
        password2 : password2,
        title:  'Please Login'
    })
     } else {
        //validation passed
       User.findOne({email : email}).exec((err,user)=>{
        console.log(user);   
        if(user) {
            errors.push({msg: 'email already registered'});
            res.render(Helper.findView('register'),{errors,name,email,password,password2,title: 'Please Login'})  
           } else {
            const newUser = new User({
                name : name,
                email : email,
                password : password
            });
    
            //hash password
            bcrypt.genSalt(10,(err,salt)=> 
            bcrypt.hash(newUser.password,salt,
                (err,hash)=> {
                    if(err) throw err;
                        //save pass to hash
                        newUser.password = hash;
                    //save user
                    newUser.save()
                    .then((value)=>{
                        console.log(value)
                        req.flash('success_msg','You have now registered!');
                        res.redirect('/users/login');
                    })
                    .catch(value=> console.log(value));
                      
                }));
             }
       })
    }
    })



//logout
router.get('/logout',userController.userLogOut);



//dashboard
router.get('/dashboard',ensureAuthenticated,userController.userDashboard);

module.exports  = router;
