// users/controllers/userController.js
const Helper = require("../helper/data");
const passport = require('passport');
const User = require("../models/users");
const bcrypt = require('bcrypt');

/*
* User  Dashboard
*/
module.exports.userDashboard = (req,res)=>{
    res.render(Helper.findView('dashboard'),{
        user: req.user,
        title: 'My Account'
    });
};



/*
* Login form
*/
module.exports.userLoginPage = (req,res)=>{

    if(req.isAuthenticated()) {
        res.redirect('/users/dashboard'); 
    }
    res.render(Helper.findView('login'), { title: 'Please Login'});
};


/*
* Login post
*/
module.exports.userLoginPost = (req,res,next)=>{
passport.authenticate('local',{
    successRedirect : '/users/dashboard',
    failureRedirect: '/users/login',
    failureFlash : true
})(req,res,next)
}




/*
* Register form
*/
module.exports.userRegisterPage = (req,res)=>{
     if(req.isAuthenticated()) {
        res.redirect('/users/dashboard'); 
    }
    res.render(Helper.findView('register'), { title: 'Please Register'})
};




/*
* Register Post
*/
module.exports.userRegisterPost = (req,res)=>{
    const {name,email, password, password2} = req.body;
    let errors = [];
    if(!name || !email || !password || !password2) {
        errors.push({msg : "Please fill in all fields"})
    }
    //check if match
    if(password !== password2) {
        errors.push({msg : "passwords dont match"});
    }
    //check if password is more than 6 characters
    if(password.length < 6 ) {
        errors.push({msg : 'password atleast 6 characters'})
    }
    if(errors.length > 0 ) {
    res.render(Helper.findView('register'), {
        errors : errors,
        name : name,
        email : email,
        password : password,
        password2 : password2,
        title:  'Please Login'
    })
     } else {
        //validation passed
       User.findOne({email : email}).exec((err,user)=>{
        if(user) {
            errors.push({msg: 'email already registered'});
            res.render(Helper.findView('register'),{errors,name,email,password,password2,title: 'Please Login'})  
           } else {
            const newUser = new User({
                name : name,
                email : email,
                password : password
            });
    
            //hash password
            bcrypt.genSalt(10,(err,salt)=> 
            bcrypt.hash(newUser.password,salt,
                (err,hash)=> {
                    if(err) throw err;
                        //save pass to hash
                        newUser.password = hash;
                    //save user
                    newUser.save()
                    .then((value)=>{
                        req.flash('success_msg','You have now registered!');
                        res.redirect('/users/login');
                    })
                    .catch(value=> console.log(value));
                }));
             }
       })
    }
    };





/*
* Log out User
*/
module.exports.userLogOut = (req,res)=>{
    req.logout();
    req.flash('success_msg','Now logged out');
    res.redirect('/users/login'); 
};
